File: rust/pr031.rs
// Problem 31
//
// 22 November 2002
//
//
// In England the currency is made up of pound, -L-, and pence, p, and there
// are eight coins in general circulation:
//
//     1p, 2p, 5p, 10p, 20p, 50p, -L-1 (100p) and -L-2 (200p).
//
// It is possible to make -L-2 in the following way:
//
//     1x-L-1 + 1x50p + 2x20p + 1x5p + 1x2p + 3x1p
//
// How many different ways can -L-2 be made using any number of coins?
//
// 73682

// Using a list avoids a lot of vector copies, and makes this about 10x faster.

// Needed until list is converted to use Rc or Gc.
#[feature(managed_boxes)]

extern mod extra;
use extra::list;
use extra::list::List;

fn main() {
    println(format!("{}", rways(200, allCoins())));
}

fn rways(remaining: uint, coins: @List<uint>) -> uint {
    if list::is_empty(coins) {
        if remaining == 0 { 1 } else { 0 }
    } else {
        let coin = list::head(coins);
        let others = list::tail(coins);

        // TODO: Is there a range-like operator that can decrease?
        let mut total = 0;
        let mut r = remaining as int;
        while r >= 0 {
            let r2 = r as uint;
            total += rways(r2, others);

            r -= coin as int;
        }
        total
    }
}

fn allCoins() -> @List<uint> {
    // Best performance puts the largest coins first.
    list::from_vec([200u, 100, 50, 20, 10, 5, 2, 1])
}

File: python3/pr031.py
#! /usr/bin/env python3

"""
Problem 31

22 November 2002


In England the currency is made up of pound, £, and pence, p, and there
are eight coins in general circulation:

    1p, 2p, 5p, 10p, 20p, 50p, £1 (100p) and £2 (200p).

It is possible to make £2 in the following way:

    1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p

How many different ways can £2 be made using any number of coins?

73682
"""

def main():
    print(rways(200, all_coins))

# Best performance puts the largest coins first.
all_coins = [200, 100, 50, 20, 10, 5, 2, 1]

def rways(remaining, coins):
    if len(coins) == 0:
        if remaining == 0:
            return 1
        else:
            return 0

    coin = coins[0]
    others = coins[1:]

    total = 0
    for r in range(remaining, -1, -coin):
        total += rways(r, others)

    return total

if __name__ == '__main__':
    main()
